epoxy-adapter/src/main/java/com/airbnb/epoxy/KeyedListeners.kt

```diff
@@ -0,0 +1,163 @@
+
+inline fun <reified V : Any, reified T : EpoxyModel<V>, Listener> T.keyedListener(
+    key: Any = this,
+    listener: Listener
+) = KeyedListener(key, listener)
+
+inline fun <reified V : Any, reified T : EpoxyModel<V>> T.keyedLongClickListener(
+    crossinline listener: ClickContext<View, V, T>.() -> Boolean
+) = KeyedLongClickListener(this, View.OnLongClickListener { v ->
+    val context = ClickContext.from<View, V, T>(v) ?: return@OnLongClickListener false
+    return@OnLongClickListener context.listener()
+})
+
+class KeyedLongClickListener(
+    key: Any,
+    listener: View.OnLongClickListener
+) : KeyedListener<View.OnLongClickListener>(key, listener), View.OnLongClickListener {
+
+    override fun onLongClick(v: View) = listener.onLongClick(v)
+}
+
+inline fun <reified V : Any, reified T : EpoxyModel<V>> T.keyedCheckedChangeListener(
+    crossinline listener: ClickContext<CompoundButton, V, T>.(isChecked: Boolean) -> Unit
+) = KeyedCheckedChangedListener(this, CompoundButton.OnCheckedChangeListener { v, isChecked ->
+    ClickContext.from<CompoundButton, V, T>(v)?.apply {
+        listener(isChecked)
+    }
+})
+
+class KeyedCheckedChangedListener(
+    key: Any,
+    listener: CompoundButton.OnCheckedChangeListener
+) : KeyedListener<CompoundButton.OnCheckedChangeListener>(key, listener),
+    CompoundButton.OnCheckedChangeListener {
+
+    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
+        listener.onCheckedChanged(buttonView, isChecked)
+    }
+}
+
+inline fun <reified V : Any, reified T : EpoxyModel<V>> T.keyedClickListener(
+    crossinline listener: ClickContext<View, V, T>.() -> Unit
+) = KeyedClickListener(this, View.OnClickListener { v ->
+    ClickContext.from<View, V, T>(v)?.apply(listener)
+})
+
+class KeyedClickListener(
+    key: Any,
+    listener: View.OnClickListener
+) : KeyedListener<View.OnClickListener>(key, listener), View.OnClickListener {
+
+    override fun onClick(v: View) = listener.onClick(v)
+}
+
+open class KeyedListener<Listener>(private val key: Any, val listener: Listener) {
+
+    // Only include the key, and not the listener, in equals/hashcode
+    override fun equals(other: Any?): Boolean {
+        val model = key as? EpoxyModel<*>
+        if (model?.inKeyedListenerEquals == true) return true
+        model?.inKeyedListenerEquals = true
+
+
+        if (this === other) return true
+        if (other !is KeyedListener<*>) return false
+        val isEqual = key == other.key
+
+        model?.inKeyedListenerEquals = false
+        return isEqual
+    }
+
+    override fun hashCode(): Int {
+        val model = key as? EpoxyModel<*>
+        if (model?.inKeyedListenerEquals == true) return 0
+        model?.inKeyedListenerEquals = true
+
+        val result = key.hashCode()
+
+        model?.inKeyedListenerEquals = false
+        return result
+    }
+}
+
+data class ClickContext<V : View, P : Any, T : EpoxyModel<P>>(
```

Comment:
The name "ClickContext" confused me a bit at first, until I understood that it's not directly related to clicks. It's just context about a view, and could be used at any time, including when non-click events happen, right?